### ISSUE TITLE
Unexpected TypeScript error with interface and generics

### ISSUE THREAD

**Issue by brunouber** [microsoft/TypeScript#55828]
### Acknowledgement

- [X] I acknowledge that issues using this template may be closed without further explanation at the maintainer's discretion.

### Comment

Not sure this is a bug in TypeScript, or my expectations are not right.
Consider these two types:

```ts
interface A {
    details?: any;
    id: string;
    value: string;
}

interface B extends A {
    details: {
        description: string;
    };
    id: 'b';
}
```

`B`is a specialization of `A` - and I expect to be able to pass `B` everywhere an `A` is accepted.

Indeed the example below works fine.

```ts
function fn(v: A) {
    return v;
}
const b: B = { details: { description: '' }, id: 'b', value: 'b' };
fn(b);
```

Below another example, a bit more complex using generics, that still works fine.

```ts
type RecordTest<T extends A = A> = Record<string, T>;
const a1: RecordTest<A> = { a: { id: 'a', value: 'a' } };
const b1: RecordTest<B> = { b: { details: { description: '' }, id: 'b', value: 'b' } };

export const test1: Record<string, RecordTest> = {
    a: a1,
    b: b1
};
```

Finally, here's where my expectation and reality diverge.
This example is not much different then previous one, except that now the generic is used as argument of a function (instead of value of an object.)

```ts
type FuncTest<T extends A = A> = (i: T) => void;
const fn1: FuncTest<A> = (v: A) => v;
const fn2: FuncTest<B> = (v: B) => v;
export const test2: Record<string, FuncTest> = {
    a: fn1,
    b: fn2
};
```

In this case I get the following error:

<img width="520" alt="image" src="https://github.com/microsoft/TypeScript/assets/55399398/600e3e35-92d3-4dc4-8e93-3145df20bcf1">

```
Type 'FuncTest<B>' is not assignable to type 'FuncTest<A>'.
  Type 'A' is not assignable to type 'B'.
    Property 'details' is optional in type 'A' but required in type 'B'.ts(2322)
```

Is this a bug, or what I'm missing?

**Comment by MartinJohns** [comment]
You're trying to assign the function `fn2` typed `FuncTest<B>` to a property typed `FuncTest<A>` (each property of `test2` has this type). The function will be called with arguments of type `A` (where `details` is optional), but your function implementation requires it to be typed `B` (where `details` is required).

**Comment by brunouber** [comment]
@MartinJohns I see it now - thank you for the clear explanation.
